### PR TITLE
Enable CI build on Azure DevOps

### PR DESCRIPTION
**Pull request by mallibone**
These changes enable building Fabulous on Azure DevOps as discussed in the issue [269](https://github.com/fsprojects/Fabulous/issues/269). Tests are executed on macOS. The builds where executed on a fork: [Build Status?branchName=AzureDevops)](https://dev.azure.com/gnabber/Fabulous/_build/latest?definitionId=22&branchName=AzureDevops)

* Includes the build definition as yaml file
* Adjustments to paths due to differences in the path where .Net Core is installed on ADO

The Linux build is currently still *disabled* due to some issues with the Generator project.

### CODE REVIEW

**Comment by TimLariviere** [comment]
I'm wondering. Is there a way to download the artifacts generated?

**Comment by TimLariviere** [comment]
Also it would be awesome if we can report the test results in the "Test" tab of Azure DevOps :)

**Comment by TimLariviere** [comment]
@mallibone Yes, you can pick the one you want.
Today, when I release a new version, I only get the artifacts from AppVeyor.

I think we can settle on only one test framework.
Fabulous.Cli.Tests uses MSTest, but I think it's mainly because it was the easiest one to include, given it's more integration tests than anything else.

File: azure-devops.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+phases:
+- phase: Windows
+  displayName: Windows
+
+  condition: succeeded()
+  queue:
+    name: Hosted VS2017
+  steps:
+  - script: |
+      echo Install fake-cli & GTK#
```
Comment (TimLariviere) [inline azure-devops.yml]: I think the `&` is interpreted.
It doesn't fail the build, but an error message is shown in the build logs.
```
Install fake-cli 
'GTK#' is not recognized as an internal or external command,
```

Comment (mallibone) [reply azure-devops.yml]: good catch 👍

**Comment by mallibone** [comment]
@TimLariviere I am okay with merging 🙈 still have some things I would like to try out, but I can do that on another branch. Will rebase from master once more to solve the conflicts.

The following options are still on my table:
* Play around with the command line arguments a bit more - there is still one option that some references show which is currently not yet set.
* Migrate to VSTest - It seems that trx is supported out of the box.

I have tried extensive profanity, but so far no luck... 😉

**Comment by mallibone** [comment]
I am currently running into a [new error](https://dev.azure.com/gnabber/Fabulous/_build/results?buildId=589&view=logs&jobId=a5e52b91-c83f-5429-4a68-c246fc63a4f7&taskId=61b979ba-27c9-588c-740e-a675d85b471e&lineStart=3711&lineEnd=3712&colStart=1&colEnd=1) on the macOS Agent. It seems that `fscd` is no longer happy.

Since all the other agents (Azure, Travis & AppVeyor) are happy with the changes, I am currently assuming this must be some weird configuration diff - will have to give this a scan later today 🔍

**Comment by TimLariviere** [comment]
Might be related to https://github.com/fsprojects/Fabulous/pull/303

**Comment by mallibone** [comment]
@TimLariviere could be, but why does it not affect Travis then? 🤔

**Comment by TimLariviere** [comment]
@mallibone I'm currently working on Azure DevOps as well. (continued from where you were)
Now, only Linux fails (wrong msbuild -- can't build .NET Standard).
Changed the build pipeline to be more flexible and removed test reporting.
https://github.com/TimLariviere/Fabulous/pull/12

**Comment by mallibone** [comment]
@TimLariviere oh cool you have been busy I see - very busy when looking at the merge history 😄 I will have a look at your branch then. It seems like the `fscd` issue is solved 🙂 may I ask what the solution was?

**Comment by TimLariviere** [comment]
A good chunk of `fscd` has been exported to https://github.com/fsprojects/FSharp.Compiler.PortaCode
Most of the unit tests as well, but Paket still imports the unit test files from PortaCode.

I think somewhere in this PR, you changed the way we resolved the dlls for the unit tests due to `dotnet` packages not being in the same place (the NuGetFallbackFolder).
But this was lost with the exportation.

So Fabulous.Cli.Tests used the wrong path for dlls.
https://github.com/TimLariviere/Fabulous/blob/51146eee1159132f51c1fd59b4b6495948c14948/tests/Fabulous.Cli.Tests/TestHelpers.fs#L173-L184

**Comment by mallibone** [comment]
> 
> 
> @mallibone I'm currently working on Azure DevOps as well. (continued from where you were)
> Now, only Linux fails (wrong msbuild -- can't build .NET Standard).
> Changed the build pipeline to be more flexible and removed test reporting.
> [TimLariviere#12](https://github.com/TimLariviere/Fabulous/pull/12)

The issue is that F# is not pre-installed on the Linux build agents - now, now, cool down @TimLariviere  & @dsyme, I am sure they have no idea on what they are missing out by not having this installed by default 🙃 (sorry could not resist this part 😇)

The latest commit(s) contain the changes from your Branch Timothé. I like the splitting into jobs - but I'm guessing you do not want to run all of them on every commit?

Further Linux now only builds GTK no Android, iOS, macOS, UWP and WPF though it would be possible to compile for [Android on Linux](https://beyondvelocity.blog/2018/11/29/xamarin-android-on-linu/).

**Comment by TimLariviere** [comment]
Good job finding that F# was missing from the build agent!

>The latest commit(s) contain the changes from your Branch Timothé. I like the splitting into jobs - but I'm guessing you do not want to run all of them on every commit?

I'm divided on this question.
We definitely need 2 separate jobs : one fast (build/test = ~5min), one complete (build/test/samples/nupkg/templates = ~25min)

Having the 3 platforms for every PRs would be awesome.
Immediate feedback if something goes wrong on a platform.
But I think it's a bit too much for pretty much all PRs.

So maybe only do 1 fast and 1 complete for PR; and 3 complete for "master"?

**Comment by mallibone** [comment]
Well sometimes one get's lucky with Google 🙂

Since the jobs run in parallel the longest job make the duration until the feedback is here. That being said, I am a fan of fast feedback in my daily job. This being an OSS project a change could result in breaking i.e. the generated project. If we only run these tests after merging on main - well the core maintainers will have to deal with a broken main branch asap.

That being said we could make the build smarter by defining multiple yaml files for ADO and depending on where the change has happened we can expand the build. I.e. if only the core of Fabulous has been edited. Or we only enable the fast builds automatically and the long runs can be triggered [manually]( https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops#trigger-builds-using-github-pull-request-comments) when reviewing.
The latter would be quicker implemented and probably less fragile should a folder path change. For the "intelligent" build definitions, it would be good to know which paths contain items that are only tested in the 25min build.

**Comment by TimLariviere** [comment]
>If we only run these tests after merging on main - well the core maintainers will have to deal with a broken main branch asap.

It's more or less what we do today. We build fast against Windows/AppVeyor (no samples/templates tests) and slow with macOS/Travis.
If Travis succeeds, the probability the same slow build on Windows and Linux will fail is very low.

>That being said we could make the build smarter by defining multiple yaml files for ADO and depending on where the change has happened we can expand the build. I.e. if only the core of Fabulous has been edited

That's what I thought at first, but if only the core is modified, it might impact both the samples and the templates so it will need a complete build as well.

>Or we only enable the fast builds automatically and the long runs can be triggered [manually](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops#trigger-builds-using-github-pull-request-comments) when reviewing.

That's really interesting!
Fast feedback, initial reviewing by something and then a complete build, while minimizing the number of automatic builds launched.
I think it's the best compromise.

**Comment by mallibone** [comment]
Okay so there are now two definitions for Azure:

- PR: Only executes the build fake config on Linux, macOS and Windows
- CI: Is only enabled for the master branch. Runs the full fake config on Linux, macOS and Windows. 

The PR build definition is named `build`, and the CI build definition `full`. So following the [docs ](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops#trigger-builds-using-github-pull-request-comments) the following command should execute the full CI build: `/AzurePipelines run full`.

**Comment by TimLariviere** [comment]
@mallibone Seems like there's a few issues:
- The `full` build runs by default on PRs.
- `/AzurePipelines` commands don't seem to have an effect?

See https://github.com/TimLariviere/Fabulous/pull/13 and https://github.com/TimLariviere/Fabulous/pull/14

**Comment by mallibone** [comment]
@TimLariviere sorry about the `full` build on every pr... fixed that one with the last commit. I will have to look into the commands.

**Comment by mallibone** [comment]
@TimLariviere Currently chating with Damien Brady to get this to work with comments but so far I haven't yet managed to get it running... Should we go with a PR build containing one long run (i.e. on macOS) and a CI build on master running all three platforms on full? Or did you have more success on one of your test builds and it some weird configuration on my side?

**Comment by TimLariviere** [comment]
@mallibone I do get something to work, but it's pretty unreliable...
PR with build-only : https://github.com/TimLariviere/Fabulous/pull/19
PR with comment to run full build : https://github.com/TimLariviere/Fabulous/pull/21

I noticed 2 things :
- For multiple simultaneous PRs, Azure DevOps will run the PR Build only for the first one. The others are forever ignored. Once the first build is done, a new PR will be granted the privilege to run a new build...
- When asking for a full build, the Azure Pipeline bot says the trigger exclude this branch (which is not the case). It then starts 3 full builds, cancels 2 of them and the last one runs successfully.
But Github only reports the first canceled build...

**Comment by TimLariviere** [comment]
This for example : https://github.com/TimLariviere/Fabulous/pull/23
Another PR was already running, so Azure Pipelines decided there was no need to run the build for this PR.
And when forced to, Azure Pipelines said there was no build for this PR...

**Comment by mallibone** [comment]
@TimLariviere hmm okay, that is interesting... 🤔 I just triggered some tests myself [branch1](https://github.com/mallibone/Fabulous/pull/6) and [branch2](https://github.com/mallibone/Fabulous/pull/5) - but I will forward your feedback. The error message is a known bug I was told, and they are working on it.

I can't reproduce the not firing trigger - there is a limit to 10 build jobs in parallel, but I would have thought they would just start to queue up. I will run some tests on my own and will try to forward any buggy results. But your findings do not sound all that great.

**Comment by TimLariviere** [comment]
Replaced by #413

**Pull request closed without merging**